### DISCUSSION TITLE
New Style of Cooperative Gameplay

### DISCUSSION THREAD

**Discussion by osuWhite** [ppy/osu#24672] [Ideas]
I wouldn't be surprised if this was suggested previously, but the few minutes I spent searching the github for previous suggestions found nothing that matched this so I'll go ahead and mention it.

I've been playing around with trying to get a true "cooperative" beatmap. Instead of TAG maps, which involve the player playing for a bit, then having to wait, I wanted to see a kind of map that allowed for a truly collaborative gameplay experience without the gaps in the action. 

Songs are comprised of multiple layers of melodies and instrumentals, and why can't maps. So, I tried to make a map which combined two distinct maps that each followed a distinctive layer of the music. Here is one of these examples: https://streamable.com/nuzena https://cdn.discordapp.com/attachments/961473695690412082/1145899937490554953/Deco27_-_831.osz

In that map I sketched, the left side follows the vocals, and the right follows the piano. This enables the mapper to create really engaging gameplay experiences for two or more players, and each time the player plays the map, simply by changing which side they play, they can have an entirely different experience. 

However, there's obvious and significant limitations with this. For one, the current way to do this is via 2B, and there's a lot of known issues with that. Ideally, I think, these could be modular difficulties, where the players can play one side individually, but also as a team that plays both sides. Ideally as well would be if the other player's cursor was visible, as that would allow the mappers to create interesting situations by encouraging both players' cursors to interact. Most important would be the need for the comboing and gameplay experience to work, however. 

While it is possible to successfully pass the map (HP0, OD10) without NoFail, currently the game has really strange interactions when playing this kind of map. It seems that if one player clicks, it _sometimes_ registers that click for the other player as well. This can result in double clicking, sliders getting released despite the player playing it properly, misses/100s/50s when the player played the map properly. Seemingly, it also causes one or both players to note lock despite a slow song like this being pretty impossible to notelock otherwise. The client cannot score combos correctly either, with the game ending combos despite that player full-comboing on their client (I assume because it registers the "partner" as missing most of their notes due to the turn overlap).  

I have tried several times to attempt to find a logic to when the game registers a "miss" despite the player hitting it properly in order to find a potential loophole that would support this kind of gameplay, but have been unable to.

I don't know how complex of a feature this would be, but I could see it being certainly possible. Perhaps, instead of having 2B support, it could be that modular design where each player plays "half" of a map and then their half is streamed to the other player?

Curious what you think!

**Comment by peppy** [comment]
Has been suggested before yep! As early as 200x.

As a starting point we hope to allow players to play different difficulties of the same beatmap in coop.

**Comment by osuWhite** [comment]
Very nice.
